How to manipulate “Difference” several times?

How could I accomplish Difference(Geo0, Geo1, Geo2)? Directly, I tried Difference(Geo0, Geo1, Geo2), and it does not work for this manipulation.    Actually, I tried GeoTemp = Difference(Geo0, Geo1), Geo = Difference(GeoTemp, Geo2), it still failed.
I check out the code of Difference, multi-operation seems not to be supported.   Does this function only support the Difference for two geometries? Is there any other option for several times of Difference?
Thank you!

I tried the way you mentioned. It seems dmsh.Difference works well, but the dmsh.generate() failed. The error message and my code are listed below:
(1) code
r = dmsh.Rectangle(60, 330, 380, 650) 
geo1 = dmsh.Union( [dmsh.Circle([194, 543], 51), dmsh.Rectangle(143, 245, 440, 543)] )
geo = dmsh.Difference(r, geo1)
geo.plot( )
show( )

np.random.seed(0)
x, cells = dmsh.generate( geo, 10, tol=1.0e-5 )  ###  lower tol to reduce the computation cost

plt.triplot(x[:,0], x[:,1], cells)
plt.show()

(2) error mesage
x, cells = dmsh.generate( geo, 10, tol=1.0e-5 )  ### lower tol to reduce the computation cost
  File "C:\Users\User\AppData\Roaming\Python\Python36\site-packages\dmsh\main.py", line 163, in generate
    pts[is_outside] = geo.boundary_step(pts[is_outside].T).T
  File "C:\Users\User\AppData\Roaming\Python\Python36\site-packages\dmsh\geometry\difference.py", line 48, in boundary_step
    assert k <= max_steps, "Exceeded maximum number of boundary steps."
AssertionError: Exceeded maximum number of boundary steps.

Thank you for help!

--- a/README.md
+++ b/README.md
@@ -7,7 +7,6 @@
 [![codecov](https://img.shields.io/codecov/c/github/nschloe/dmsh.svg?style=flat-square)](https://codecov.io/gh/nschloe/dmsh)
 [![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg?style=flat-square)](https://github.com/psf/black)
 [![PyPi Version](https://img.shields.io/pypi/v/dmsh.svg?style=flat-square)](https://pypi.org/project/dmsh)
-[![Debian CI](https://badges.debian.net/badges/debian/testing/python3-dmsh/version.svg?style=flat-square)](https://tracker.debian.org/pkg/python-dmsh)
 [![GitHub stars](https://img.shields.io/github/stars/nschloe/dmsh.svg?style=flat-square&logo=github&label=Stars&logoColor=white)](https://github.com/nschloe/dmsh)
 [![PyPi downloads](https://img.shields.io/pypi/dm/dmsh.svg?style=flat-square)](https://pypistats.org/packages/dmsh)
 

--- a/dmsh/__about__.py
+++ b/dmsh/__about__.py
@@ -2,5 +2,5 @@ __author__ = "Nico Schlömer"
 __email__ = "nico.schloemer@gmail.com"
 __copyright__ = "Copyright (c) 2018-2019, {} <{}>".format(__author__, __email__)
 __license__ = "License :: OSI Approved :: MIT License"
-__version__ = "0.1.4"
+__version__ = "0.1.5"
 __status__ = "Development Status :: 4 - Beta"

--- a/dmsh/geometry/union.py
+++ b/dmsh/geometry/union.py
@@ -30,14 +30,43 @@ class Union(Geometry):
     def dist(self, x):
         return numpy.min([geo.dist(x) for geo in self.geometries], axis=0)
 
-    def boundary_step(self, x):
-        # Step to the geometry with the smallest distance. The new point is not always
-        # on the domain boundary.
+    def boundary_step(self, x, tol=1.0e-5, max_steps=10):
+        # This stepper uses a heuristic that doesn't always land on the closet point in
+        # the boundary. It's good enough for now though.
+        x = numpy.asarray(x)
+        assert x.shape[0] == 2
+
+        is_one_dimensional = False
+        if len(x.shape) == 1:
+            is_one_dimensional = True
+            x = x.reshape(-1, 1)
+
         out = x.copy()
-        alpha = numpy.array([geo.dist(x) for geo in self.geometries])
-        idx = numpy.argmin(numpy.abs(alpha), axis=0)
-        for k, geo in enumerate(self.geometries):
-            j = idx == k
-            if numpy.any(j):
-                out[:, j] = geo.boundary_step(x[:, j])
+
+        # At the end of this, min(dist(domains)) must be 0 for all x, so each x either
+        # sits outside or on the boundary of each domain. Idea: Use the domain with the
+        # min value for stepping, and do that until all x are on the boundary.
+        dists = numpy.array([geo.dist(out) for geo in self.geometries])
+        dist = numpy.min(dists, axis=0)
+        needs_stepping = (dist < -tol) | (tol < dist)
+
+        k = 0
+        while numpy.any(needs_stepping):
+            assert k <= max_steps, "Too many union boundary steps"
+            # Step to the boundary of the domain with the smallest value. (Either it's
+            # far inside or domain or, if outside of the union, the closest to a
+            # boundary.)
+            idx = numpy.argmin(dists[:, needs_stepping], axis=0)
+            for i, geo in enumerate(self.geometries):
+                j = (idx == i)
+                if numpy.any(j):
+                    out[:, j] = geo.boundary_step(out[:, j])
+
+            dists = numpy.array([geo.dist(out) for geo in self.geometries])
+            dist = numpy.min(dists, axis=0)
+            needs_stepping = (dist < -tol) | (tol < dist)
+            k += 1
+
+        if is_one_dimensional:
+            out = out.reshape(-1)
         return out

--- a/dmsh/geometry/union.py
+++ b/dmsh/geometry/union.py
@@ -58,7 +58,7 @@ class Union(Geometry):
             # boundary.)
             idx = numpy.argmin(dists[:, needs_stepping], axis=0)
             for i, geo in enumerate(self.geometries):
-                j = (idx == i)
+                j = idx == i
                 if numpy.any(j):
                     out[:, j] = geo.boundary_step(out[:, j])
 

--- a/test/helpers.py
+++ b/test/helpers.py
@@ -6,9 +6,9 @@ def assert_equality(a, b, tol):
     b = numpy.asarray(b)
     fmt_a = ", ".join(["{:.16e}"] * len(a))
     fmt_b = ", ".join(["{:.16e}"] * len(b))
-    assert numpy.all(
-        numpy.abs(a - b) < tol
-    ), ("[{}]\n[{}]".format(fmt_a, fmt_b)).format(*a, *b)
+    assert numpy.all(numpy.abs(a - b) < tol), (
+        "[{}]\n[{}]".format(fmt_a, fmt_b)
+    ).format(*a, *b)
 
 
 def assert_norm_equality(X, ref_norm, tol):

--- a/test/helpers.py
+++ b/test/helpers.py
@@ -1,8 +1,18 @@
 import numpy
 
 
+def assert_equality(a, b, tol):
+    a = numpy.asarray(a)
+    b = numpy.asarray(b)
+    fmt_a = ", ".join(["{:.16e}"] * len(a))
+    fmt_b = ", ".join(["{:.16e}"] * len(b))
+    assert numpy.all(
+        numpy.abs(a - b) < tol
+    ), ("[{}]\n[{}]".format(fmt_a, fmt_b)).format(*a, *b)
+
+
 def assert_norm_equality(X, ref_norm, tol):
-    ref_norm = numpy.array(ref_norm)
+    ref_norm = numpy.asarray(ref_norm)
     vals = numpy.array(
         [
             numpy.linalg.norm(X, ord=1),

--- a/test/test_union.py
+++ b/test/test_union.py
@@ -1,5 +1,20 @@
+import numpy
+
 import dmsh
-from helpers import assert_norm_equality, save
+from helpers import assert_norm_equality, save, assert_equality
+
+
+def test_boundary_step():
+    geo = dmsh.Union([dmsh.Circle([-0.5, 0.0], 1.0), dmsh.Circle([+0.5, 0.0], 1.0)])
+    geo.show()
+    a = geo.boundary_step([-0.5, 0.9])
+    assert numpy.array_equal(a, [-0.5, 1.0])
+
+    a = geo.boundary_step([-0.5, 0.6])
+    assert numpy.array_equal(a, [-0.5, 1.0])
+
+    a = geo.boundary_step([0.05, 0.05])
+    assert_equality(a, [-4.4469961425821203e-01, 9.9846976285554556e-01], 1.0e-12)
 
 
 def test_union(show=False):
@@ -14,5 +29,6 @@ def test_union(show=False):
 
 
 if __name__ == "__main__":
-    X, cells = test_union(show=False)
-    save("union.png", X, cells)
+    test_boundary_step()
+    # X, cells = test_union(show=False)
+    # save("union.png", X, cells)

--- a/test/test_union.py
+++ b/test/test_union.py
@@ -1,7 +1,7 @@
 import numpy
 
 import dmsh
-from helpers import assert_norm_equality, save, assert_equality
+from helpers import assert_equality, assert_norm_equality, save
 
 
 def test_boundary_step():
@@ -30,5 +30,5 @@ def test_union(show=False):
 
 if __name__ == "__main__":
     test_boundary_step()
-    # X, cells = test_union(show=False)
-    # save("union.png", X, cells)
+    X, cells = test_union(show=False)
+    save("union.png", X, cells)

